lib/Interpreter/Interpreter.cpp

```diff
@@ -276,44 +297,109 @@ namespace cling {
+      } else {
+        Strm << "#include \"cling/Interpreter/CValuePrinter.h\"\n"
+                "void* gCling=(void*)" << thisP << ";\n";
+      }
+    }
 
-    std::stringstream initializer;
-#ifdef _WIN32
-    // We have to use the #defined __CLING__ on windows first.
-    //FIXME: Find proper fix.
-    initializer << "#ifdef __CLING__ \n#endif\n";
+    // Intercept all atexit calls, as the Interpreter and functions will be long
+    // gone when the -native- versions invoke them.
+    if (!SyntaxOnly) {
+#if defined(__GLIBCXX__) && !defined(__APPLE__)
+      const char* LinkageCxx = "extern \"C++\"";
+      const char* Attr = LangOpts.CPlusPlus ? " throw () " : "";
+#else
+      const char* LinkageCxx = Linkage;
+      const char* Attr = "";
 #endif
 
-    initializer << "#include \"cling/Interpreter/RuntimeUniverse.h\"\n";
+      // While __dso_handle is still overriden in the JIT below,
+      // #define __dso_handle is used to mitigate the following problems:
+      //  1. Type of __dso_handle is void* making assignemnt to it legal
+      //  2. Making it void* const in cling would mean possible type mismatch
+      //  3. Cannot override void* __dso_handle in child Interpreter
+      //  4. On Unix where the symbol actually exists, __dso_handle will be
+      //     linked into the code before we the JIT can say otherwise, so:
+      //      [cling] __dso_handle // codegened __dso_handle always printed
+      //      [cling] __cxa_atexit(f, 0, __dso_handle) // seg-fault
+      //  5. Code that actually uses __dso_handle will fail as a declaration is
+      //     needed which is not possible with the macro.
+      //  6. Assuming 4 is sorted out in user code, calling __cxa_atexit through
+      //     atexit below isn't linking to the __dso_handle symbol.
+
+      Strm << "#define __dso_handle ((void*)" << thisP << ")\n";
+
+      // Use __cxa_atexit to intercept all of the following routines
+      Strm << Linkage << " int __cxa_atexit(void (*f)(void*), void*, void*);\n";
+
+      // C atexit, std::atexit
+      Strm << Linkage << " int atexit(void(*f)()) " << Attr << " { return "
+                        "__cxa_atexit((void(*)(void*))f, 0, __dso_handle); }\n";
+      Globals.push_back("atexit");
+
+      // C++ 11 at_quick_exit, std::at_quick_exit
+      if (LangOpts.CPlusPlus && LangOpts.CPlusPlus11) {
+        Strm << LinkageCxx << " int at_quick_exit(void(*f)()) " << Attr <<
+              " { return __cxa_atexit((void(*)(void*))f, 0, __dso_handle); }\n";
+        Globals.push_back(LinkageCxx == Linkage ?
+                          "at_quick_exit" : "_Z13at_quick_exitPFvvE");
+      }
+
+#if defined(LLVM_ON_WIN32)
+      // Windows specific: _onexit, _onexit_m, __dllonexit
+ #if !defined(_M_CEE)
+      const char* Spec = "__cdecl";
+ #else
+      const char* Spec = "__clrcall";
+ #endif
+      Strm << Linkage << " " << Spec << " int (*__dllonexit("
+           << "int (" << Spec << " *f)(void**, void**), void**, void**))"
```

Comment:
I had this discussion with @vgvassilev a bit...and while I agree the strings aren't great, I'm generally against putting it into _RuntimeUniverse.h_ for the following reasons:

1. Having it there means anytime the file is missing/corrupt means destructors will **never** be run (vs the current behavior of only printing not working). i.e. one missing file and C++ conformance is gone!
2. Not including _RuntimeUniverse.h_ can be beneficial for debugging type declarations as it keeps the AST _purer_.
3. I'm not sure how a resolution of **__dso_handle** would work when the functions are defined in a PCH and that PCH is loaded before **Interpreter::Initialize** is called.
4. It will be ugly as well requiring at least 3 additional macros.